### ISSUE TITLE
Support of "Connection close" headers for HTTPS through proxy

### ISSUE THREAD

**Issue by axel3rd** [square/okhttp#2426]
Hi, 

This issue could be linked to #1517, but is here specific to https through proxy (So I allow myself to open a new one).

**Synthesis** : The client seems not respect `Connection: close` header come from a proxy when no preemptive authentication on a HTTPS connection.

Consider this snippet code (simple HTTP GET through a proxy) :
```
        Authenticator.setDefault(new Authenticator() {
            protected PasswordAuthentication getPasswordAuthentication() {
                return new PasswordAuthentication(USER, PASSWORD.toCharArray());
            }
        });

        Proxy PROXY = new Proxy(Type.HTTP, new InetSocketAddress(PROXY_URL, 80));
        OkHttpClient okHttpClient = new OkHttpClient();
        okHttpClient.setProxy(PROXY);
        Request request = new Request.Builder().url("http://www.google.org/").build();
        Response response = okHttpClient.newCall(request).execute();
        System.out.println(response.code());
        response.body().close();
```
The call works fine and the network trace (wireshark) is : 
```
GET http://www.google.org/ HTTP/1.1
Host: www.google.org
Connection: Keep-Alive
Accept-Encoding: gzip
User-Agent: okhttp/2.6.0

HTTP/1.1 407 Proxy Authentication Required
Mime-Version: 1.0
Date: Tue, 22 Mar 2016 19:26:29 CET
Content-Type: text/html
Proxy-Authenticate: Basic realm="[deleted]"
Connection: close
Proxy-Connection: close
Content-Length: 1723

[HTML content]
```
And **in a second TCP stream** : 
```
GET http://www.google.org/ HTTP/1.1
Proxy-Authorization: Basic [deleted]=
Host: www.google.org
Connection: Keep-Alive
Accept-Encoding: gzip
User-Agent: okhttp/2.6.0

HTTP/1.1 200 OK
[correct headers]
```

Now consider the same snippet code on HTTPS => **https://www.google.org/**, the call fails with Java stackstrace : 
```
java.io.IOException: unexpected end of stream on Connection{www.google.org:443, proxy=HTTP @ [deleted] hostAddress=[deleted] cipherSuite=none protocol=null} (recycle count=0)
	at com.squareup.okhttp.internal.http.HttpConnection.readResponse(HttpConnection.java:211)
 [...]
Caused by: java.io.EOFException: \n not found: size=0 content=...
	at okio.RealBufferedSource.readUtf8LineStrict(RealBufferedSource.java:201)
	at com.squareup.okhttp.internal.http.HttpConnection.readResponse(HttpConnection.java:191)
	... 37 more
```

The corresponding network trace is : 
```
CONNECT www.google.org:443 HTTP/1.1
Host: www.google.org
Proxy-Connection: Keep-Alive
User-Agent: okhttp/2.6.0

HTTP/1.1 407 Proxy Authentication Required
Mime-Version: 1.0
Date: Tue, 22 Mar 2016 19:31:25 CET
Content-Type: text/html
Proxy-Authenticate: Basic realm="[deleted]"
Connection: close
Proxy-Connection: close
Content-Length: 1723

[HTML content]

CONNECT www.google.org:443 HTTP/1.1
Host: www.google.org
Proxy-Connection: Keep-Alive
User-Agent: okhttp/2.6.0
Proxy-Authorization: Basic [deleted]=
```
The second ``CONNECT`` is in the **same TCP session**, but the proxy ask a new TCP session with headers ``Connection: close`` / ``Proxy-Connection: close`` and has close the connection.
=> No answer to the second ``CONNECT``, and no content got.

The below unit test *reproduces* the problem (Java exception), but with some limitations ... I didn't found how to simulate different TCP sessions with ``MockResponse`` => proxy server is stop violently.
[ProxyHttpsTest.java.txt](https://github.com/square/okhttp/files/184739/ProxyHttpsTest.java.txt)

Reproduced with v2.6.0 / 2.7.5 / 3.2.0.
Best regards.

**Comment by axel3rd** [comment]
@swankjesse : Thanks for the unit test, very usefull for provide a PR  :smiley:

About the problem, this relates the $14.10 of [RFC HTTP 2616](https://www.ietf.org/rfc/rfc2616.txt) : 
```
   in either the request or the response header fields indicates that
   the connection SHOULD NOT be considered `persistent' (section 8.1)
   after the current request/response is complete.
```

**Comment by dave-r12** [comment]
Closed with #2467.
